fml/task_source.h

```diff
@@ -0,0 +1,61 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FML_TASK_SOURCE_H_
+#define FLUTTER_FML_TASK_SOURCE_H_
+
+#include "flutter/fml/delayed_task.h"
+#include "flutter/fml/task_queue_id.h"
+
+namespace fml {
+
+class MessageLoopTaskQueues;
+
+enum TaskSourceGrade {
+  kPrimary,
+  kSecondary,
+};
+
+class TaskSource {
+ public:
+  struct TopTask {
+    TaskQueueId task_queue_id;
+    TaskSourceGrade task_source_grade;
+    DelayedTask task;
+  };
+
+  explicit TaskSource(TaskQueueId task_queue_id);
+
+  ~TaskSource();
+
+  void Clear();
```

Comment:
It isn't clear from reading the interface how and when to use `Clear` (and correctly). For instance, if the tasks are dropped immediately, won't tasks that release some resource (like a mutex) be dropped on the floor.

Reply:
Calling this `ShutDown` to make it more explicit that this shuts down the task source and drops all the tasks. This is currently only called when a message loop is shutting down. Your question about not releasing a resource is still valid though. Given that this problem exists even today when a task heap is shut down, I'm going to file an issue to investigate this separately.